Sources/PackageGraph/ModulesGraph+Loading.swift

```diff
@@ -428,8 +428,8 @@ private func createResolvedPackages(
     )
     try dupProductsChecker.run(lookupByProductIDs: moduleAliasingUsed, observabilityScope: observabilityScope)
 
-    // The set of all target names.
-    var allTargetNames = Set<String>()
+    // The array of all target names.
+    let allTargetNames = packageBuilders.reduce([]) { $0 + $1.targets }.map(\.target.name)
 
     // Track if multiple targets are found with the same name.
     var foundDuplicateTarget = false
```

Comment:
How about pre-computing `foundDuplicateTarget` here? Then we can avoid `filter {}.count` O(N) operation inside hot loop.

```suggestion
    // The set of all target names.
    var allTargetNames = Set<String>()

    // Track if multiple targets are found with the same name.
    var foundDuplicateTarget = false

    for packageBuilder in packageBuilders {
        for target in packageBuilder.targets {
            // Record if we see a duplicate target.
            foundDuplicateTarget = foundDuplicateTarget || !allTargetNames.insert(target.target.name).inserted
        }
    }
```